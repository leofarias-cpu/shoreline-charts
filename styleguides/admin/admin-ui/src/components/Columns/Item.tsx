import React, { ReactNode } from 'react'
import {
  StyleProp,
  ResponsiveValue,
  useResponsiveValue,
  useClassName,
} from '@vtex/admin-ui-system'

import { Overridable } from '../../types'
import { useColumnsContext } from './context'

export function ColumnsItem(props: ColumnsItemProps) {
  const { units, offset = 'none', styleOverrides, ...layoutProps } = props
  const { spacing } = useColumnsContext()
  const responsiveUnits = useResponsiveValue(units)
  const responsiveOffset = useResponsiveValue(offset)

  const margin = {
    left: {
      marginLeft: 'auto',
    },
    right: {
      marginRight: 'auto',
    },
    both: {
      marginLeft: 'auto',
      marginRight: 'auto',
    },
    none: {},
  }[responsiveOffset]

  const styles: StyleProp = responsiveUnits
    ? {
        flex: '0 0 auto',
        width: `${(Number(responsiveUnits) / 12) * 100}%`,
        ...margin,
        '&:not(:first-child)': {
          paddingLeft: spacing,
        },
      }
    : {
        flex: '1 1 0%',
        '&:not(:first-child)': {
          paddingLeft: spacing,
        },
        maxWidth: '100%',
      }

  const className = useClassName({
    props: { styles: { ...styles, ...styleOverrides } },
  })

  return <div className={className} {...layoutProps} />
}

export interface ColumnsItemProps extends Overridable {
  children?: ReactNode
  units?: ResponsiveValue<1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12>
  offset?: ResponsiveValue<'left' | 'right' | 'both' | 'none'>
}
