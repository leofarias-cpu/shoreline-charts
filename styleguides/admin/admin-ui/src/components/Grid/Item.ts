/* eslint-disable @typescript-eslint/no-explicit-any */
import { ElementType, ReactElement } from 'react'
import * as CSS from 'csstype'
import { createComponent, omit, pick, ResponsiveValue } from '@vtex/admin-core'

import { Primitive, PrimitiveProps } from '../Primitive'

/**
 * Grid default element
 */
const defaultElement = 'div'

export const GridItem: <E extends ElementType = typeof defaultElement>(
  props: GridItemProps<E>
) => ReactElement | null = createComponent(Primitive, useGridItem)

export function useGridItem(props: GridItemProps) {
  const propertyMap = {
    area: 'gridArea',
  }

  const { styles, ...boxProps } = props

  const cssProps = Object.keys(propertyMap)
  const cssPropsStyle = renameKeys(propertyMap, {
    ...pick(boxProps, cssProps),
  })

  return {
    styles: { ...cssPropsStyle, ...styles },
    ...omit(boxProps, cssProps),
  }
}

function renameKeys(keysMap: { [x: string]: any }, obj: { [x: string]: any }) {
  return Object.keys(obj).reduce(
    (acc, key) => ({
      ...acc,
      ...{ [keysMap[key] || key]: obj[key] },
    }),
    {}
  )
}

export interface GridItemOwnProps {
  /** Shorthand for CSS gridArea property */
  area?: ResponsiveValue<CSS.Property.GridArea>
}

export type GridItemProps<
  E extends ElementType = ElementType
> = GridItemOwnProps & PrimitiveProps<E>
