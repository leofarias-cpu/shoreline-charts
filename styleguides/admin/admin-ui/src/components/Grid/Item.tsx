/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { ElementType } from 'react'
import * as CSS from 'csstype'
import { omit, pick, ResponsiveValue } from '@vtex/admin-core'

import { Box, BoxProps } from '../Box'

/**
 * Grid default element
 */
const defaultElement = 'div'

export function GridItem<E extends ElementType = typeof defaultElement>(
  props: GridItemProps<E>
) {
  const gridItemProps = useGridItem(props)

  return <Box {...gridItemProps} />
}

export function useGridItem(props: GridItemProps) {
  const propertyMap = {
    area: 'gridArea',
  }

  const { styles, ...boxProps } = props

  const cssProps = Object.keys(propertyMap)
  const cssPropsStyle = renameKeys(propertyMap, {
    ...pick(boxProps, cssProps),
  })

  return {
    styles: { ...cssPropsStyle, ...styles },
    ...omit(boxProps, cssProps),
  }
}

function renameKeys(keysMap: { [x: string]: any }, obj: { [x: string]: any }) {
  return Object.keys(obj).reduce(
    (acc, key) => ({
      ...acc,
      ...{ [keysMap[key] || key]: obj[key] },
    }),
    {}
  )
}

export interface GridItemOwnProps {
  /** Shorthand for CSS gridArea property */
  area?: ResponsiveValue<CSS.Property.GridArea>
}

export type GridItemProps<
  E extends ElementType = ElementType
> = GridItemOwnProps & BoxProps<E>
