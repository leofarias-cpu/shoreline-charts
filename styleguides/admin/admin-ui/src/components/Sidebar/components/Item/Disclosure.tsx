import React, { forwardRef, ReactNode, Ref } from 'react'
import { merge } from '@vtex/admin-core'

import { ButtonProps, Button } from '../../../Button'
import { Tooltip } from '../../../Tooltip'

export const SidebarDisclosure = forwardRef(function SidebarDisclosure(
  props: SidebarDisclosureProps,
  ref: Ref<HTMLButtonElement>
) {
  const { icon, selected, label, csx = {}, ...buttonProps } = props

  return (
    <Tooltip
      label={label}
      placement="right"
      csx={{
        display: 'flex',
        alignItems: 'center',
        fontSize: 0,
        paddingY: 1,
        paddingX: 2,
        borderRadius: 'default',
        minHeight: 24,
        zIndex: 'over',
      }}
    >
      <Button
        ref={ref}
        variant="tertiary"
        icon={icon}
        csx={merge(
          {
            zIndex: 'sidebarDisclosure',
            backgroundColor: selected ? 'sidebar.hover' : 'unset',
            'div > svg': {
              color: selected ? 'unset' : 'sidebar.dark',
            },
            '&:hover': {
              backgroundColor: 'sidebar.hover',
              'div > svg': {
                color: 'blue',
                opacity: 1,
              },
            },
          },
          csx
        )}
        {...buttonProps}
      />
    </Tooltip>
  )
})

export interface SidebarDisclosureProps
  extends Omit<ButtonProps, 'children' | 'icon'> {
  /**
   * `icon` which the `<Sidebar.Item {...props} />` disclosure should exhibit.
   */
  icon: ReactNode
  /**
   * `label` of a `<Sidebar.Item {...props} />`. This is set as the components`
   * composite ID and its value is displayed on the tooltip.
   */
  label: string
  /**
   * This prop is shared between this component and the `<Sidebar.Item {...props} />`.
   * It indicates whether its parent component is selected or not, and is used here
   * for styling only.
   */
  selected?: boolean
}
