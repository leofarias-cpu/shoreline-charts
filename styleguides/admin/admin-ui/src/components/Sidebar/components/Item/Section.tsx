import React, { forwardRef, ReactNode, Ref } from 'react'
import { useComposite } from '../Aria'
import { Set } from '../../../Set'
import { Text } from '../../../Text'
import { SystemComponent } from '../../../../types'
import { SidebarSectionItem } from './SectionItem'
import { useItemContext } from './shared'

const _SidebarSection = forwardRef(function SidebarSection(
  props: SidebarSectionProps,
  ref: Ref<HTMLButtonElement>
) {
  const { title, children } = props
  const { state } = useItemContext()
  const compositeProps = useComposite(state)

  return (
    <Set
      ref={ref}
      spacing={0.5}
      orientation="vertical"
      csx={{
        width: '11.5rem',
        paddingBottom: 8,
      }}
      {...compositeProps}
    >
      <Text
        variant="action"
        csx={{
          color: 'dark.primary',
          fontSize: '0.6875rem',
          paddingBottom: '0.8125rem',
          paddingX: '0.75rem',
          fontSettings: 'medium',
        }}
      >
        {title}
      </Text>
      {children}
    </Set>
  )
})

export const SidebarSection = Object.assign(_SidebarSection, {
  /**
   * The last node of the sidebar tree. This is where clients
   * will interact most to perform actions.
   *
   * @example
   * ```jsx
   * import { Sidebar } from `@vtex/admin-ui`
   *
   * <Sidebar>
   *    <Sidebar.Header>
   *      <Sidebar.Item selected={someCondition} onClick={() => navigate({ to: "/promotions" })}>
   *       <Sidebar.Item.Section title={"Promotions"}>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/promotions" })}>
   *         Promotions
   *        </Sidebar.Item.Section.Item>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/discounts"})} >
   *         Discounts
   *        </Sidebar.Item.Section.Item>
   *       </Sidebar.Item.Section>
   *      </Sidebar.Item>
   *    </Sidebar.Header>
   *    <Sidebar.Footer>
   *      <Sidebar.Item selected={someCondition} onClick={() => navigate({ to: "/apps" })}>
   *       <Sidebar.Item.Section title={"Apps"}>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/apps" })}>
   *         Installed apps
   *        </Sidebar.Item.Section.Item>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/apps-store"})} >
   *         Apps store
   *        </Sidebar.Item.Section.Item>
   *       </Sidebar.Item.Section>
   *      </Sidebar.Item>
   *    </Sidebar.Footer>
   * </Sidebar>
   * ```
   */
  Item: SidebarSectionItem,
})

export interface SidebarSectionProps extends SystemComponent {
  /**
   * `title` of a section. This is what separates each item's section.
   */
  title: string
  /**
   * `chilren` should be multiple `<Sidebar.SubItem {...props} />` components.
   * Those are the items over which clients will interact in order to
   * navigate between different pages.
   */
  children?: ReactNode
}
