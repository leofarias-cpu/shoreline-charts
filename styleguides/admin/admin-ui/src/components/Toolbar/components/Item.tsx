import React, { ReactNode } from 'react'
import { ExtractHTMLAttributes } from 'reakit-utils/ts'
import { useToolbarContext } from '../context'
import { ReakitToolbarItem } from './Aria'

export function ToolbarItem(props: ToolbarItemProps) {
  const { children } = props
  const state = useToolbarContext()

  return (
    <ReakitToolbarItem {...state}>
      {(itemProps) => children(itemProps)}
    </ReakitToolbarItem>
  )
}

interface ToolbarItemProps {
  children: (itemProps: ExtractHTMLAttributes<any>) => ReactNode
}
