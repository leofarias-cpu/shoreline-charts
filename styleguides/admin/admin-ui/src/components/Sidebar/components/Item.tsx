import React, { cloneElement } from 'react'
import { SidebarDisclosure, SidebarDisclosureProps } from './Disclosure'
import { ReakitMenuItem, useMenuState } from './AriaSidebar'
import { SystemComponent } from '../../../types'
import { SidebarSectionProps } from './Section'
import { Sidebar } from '..'

export interface SidebarItemProps
  extends SidebarDisclosureProps,
    SystemComponent {
  collapsed?: boolean
  sections?: Omit<SidebarSectionProps, 'secret'>[]
}

export function SidebarItem(props: Omit<SidebarItemProps, 'secret'>) {
  const { collapsed, sections, onClick, ...baseProps } = props

  const {
    // @ts-ignore
    secret: { state: parentState },
  } = props

  const state = useMenuState({
    orientation: 'vertical',
    loop: true,
  })

  return (
    <ReakitMenuItem {...parentState}>
      {(itemProps) => (
        <>
          <SidebarDisclosure
            {...props}
            {...itemProps}
            // These prevent the sidebar to open/collapse as
            // a user hovers the mouse over the sidebar disclosure
            onMouseEnter={(event) => event.preventDefault()}
            onMouseLeave={(event) => event.preventDefault()}
            onClick={onClick}
            secret={{
              state,
            }}
          />
          {sections?.map(
            ({ title, children }, index) =>
              children.length > 0 && (
                <ReakitMenuItem {...state} key={index}>
                  {(itemProps) =>
                    cloneElement(
                      <Sidebar.Section title={title} {...baseProps}>
                        {children}
                      </Sidebar.Section>,
                      { ...itemProps, secret: { state } }
                    )
                  }
                </ReakitMenuItem>
              )
          )}
        </>
      )}
    </ReakitMenuItem>
  )
}
