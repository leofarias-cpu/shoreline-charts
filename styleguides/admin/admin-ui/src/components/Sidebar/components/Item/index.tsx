import React, { ReactNode, useMemo } from 'react'
import { HTMLAttributesWithRef } from 'reakit-utils/ts'
import { Box } from '@vtex/admin-primitives'
import { CompositeItem, useCompositeState } from '../Aria'
import { SystemComponent } from '../../../../types'
import { SidebarDisclosureProps, SidebarDisclosure } from './Disclosure'
import { SidebarSection } from './Section'
import { ItemProvider, ArrowKeys } from './shared'
import { useSidebarContext } from '../../context'
import { SCALES } from '../../consts'

function _SidebarItem(props: SidebarItemProps) {
  const { children, onClick, label, uniqueKey, icon, ...baseProps } = props

  const state = useSidebarContext()
  const selected = useMemo(() => state.isSelected(uniqueKey), [
    uniqueKey,
    state.isSelected,
  ])
  const compositeState = useCompositeState({
    orientation: 'vertical',
    loop: true,
  })

  const expandable = state.isExpandable(children)
  const translate = selected && !state.layout.reduced ? '3rem' : '-13.5rem'

  const handleSelection = () => {
    const currItem = {
      uniqueKey,
      expandable,
    }
    state.setSelectedItem(currItem)
  }

  const handleExpansion = () => {
    if (!expandable) return

    state.layout.expand()
  }

  const handleOnClick = (event: React.MouseEvent<any, MouseEvent>) => {
    handleSelection()
    handleExpansion()
    onClick?.(event)
  }

  const handleOnKeyDown = (
    event: React.KeyboardEvent<any>,
    itemProps: HTMLAttributesWithRef
  ) => {
    if (typeof itemProps.onKeyDown === 'function') {
      itemProps.onKeyDown(event)

      // Opens sidebar and focus on the first sidebar sub item
      if (event.key === ArrowKeys.Right) {
        handleSelection()
        handleExpansion()
        // We need a delay here in order to allow the object to mount,
        // otherwise there would be nothing to focus on
        setTimeout(() => {
          compositeState.first()
        }, 10)
      }
    }
  }

  const handleShowToggle = (event: React.MouseEvent<any, MouseEvent>) => {
    state.layout.showToggle()

    if (typeof baseProps.onMouseEnter === 'function') {
      baseProps.onMouseEnter(event)
    }
  }

  const handleHideToggle = (event: React.MouseEvent<any, MouseEvent>) => {
    state.layout.hideToggle()

    if (typeof baseProps.onMouseLeave === 'function') {
      baseProps.onMouseLeave(event)
    }
  }

  return (
    <CompositeItem
      {...state.composite}
      role="menuitem"
      aria-label={label}
      id={label}
    >
      {(itemProps) => (
        <>
          <SidebarDisclosure
            {...itemProps}
            icon={icon}
            selected={state.isSelected(uniqueKey)}
            label={label}
            onClick={handleOnClick}
            onKeyDown={(event) => handleOnKeyDown(event, itemProps)}
          />
          <Box
            element="ul"
            csx={{
              position: 'absolute',
              top: 0,
              maxWidth: SCALES.COLLAPSIBLE_AREA_WIDTH,
              height: '100%',
              width: '12.5rem',
              padding: '1.5rem 0.5rem',
              outline: 'none',
              overflow: 'auto',
              backgroundColor: 'transparent',
              transform: `translateX(${translate})`,
            }}
            data-testid={`${label}-ul`}
            {...baseProps}
            onMouseEnter={handleShowToggle}
            onMouseLeave={handleHideToggle}
          >
            <Box
              aria-label={`${label} menu`}
              element="li"
              csx={{
                listStyle: 'none',
              }}
            >
              <ItemProvider
                value={{
                  state: compositeState,
                  id: label,
                  selected,
                }}
              >
                {children}
              </ItemProvider>
            </Box>
          </Box>
        </>
      )}
    </CompositeItem>
  )
}

export const SidebarItem = Object.assign(_SidebarItem, {
  /**
   * Each `<Sidebar.Item.Section />` is responsible for
   * defining the scope of a section within a sidebar item.
   * It holds the `<Sidebar.Item.Section.Item />`, which
   * is the last node of the sidebar tree and where clients
   * will interact most to perform actions.
   *
   * @example
   * ```jsx
   * import { Sidebar } from `@vtex/admin-ui`
   *
   * <Sidebar>
   *    <Sidebar.Header>
   *      <Sidebar.Item selected={someCondition} onClick={() => navigate({ to: "/promotions" })}>
   *       <Sidebar.Item.Section title={"Promotions"}>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/promotions" })}>
   *         Promotions
   *        </Sidebar.Item.Section.Item>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/discounts"})} >
   *         Discounts
   *        </Sidebar.Item.Section.Item>
   *       </Sidebar.Item.Section>
   *      </Sidebar.Item>
   *    </Sidebar.Header>
   *    <Sidebar.Footer>
   *      <Sidebar.Item selected={someCondition} onClick={() => navigate({ to: "/apps" })}>
   *       <Sidebar.Item.Section title={"Apps"}>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/apps" })}>
   *         Installed apps
   *        </Sidebar.Item.Section.Item>
   *        <Sidebar.Item.Section.Item onClick={() => navigate({ to: "/apps-store"})} >
   *         Apps store
   *        </Sidebar.Item.Section.Item>
   *       </Sidebar.Item.Section>
   *      </Sidebar.Item>
   *    </Sidebar.Footer>
   * </Sidebar>
   * ```
   */
  Section: SidebarSection,
})

export interface SidebarItemProps
  extends SidebarDisclosureProps,
    SystemComponent {
  children?: ReactNode
  /**
   * `label` is set as the components` composite ID and
   * its value is displayed on its disclosure's tooltip.
   */
  label: string
  /**
   * key that defines your item as unique
   */
  uniqueKey: string | number
}
