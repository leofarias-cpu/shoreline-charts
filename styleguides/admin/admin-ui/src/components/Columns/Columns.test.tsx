import React from 'react'
import { render } from '@testing-library/react'
import { axe } from 'jest-axe'

import { Columns } from './index'
import { ThemeProvider } from '@vtex/admin-core'
import { ColumnsItem } from './Item'

describe('Columns tests', () => {
  beforeEach(() => {
    /**
     * ðŸš§ Workaround for window.match media
     * @see https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
     */
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    })
  })

  it('should have overridable styles', () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <Columns data-testid="columns" styleOverrides={{ bg: 'coral' }}>
          <ColumnsItem data-testid="item" styleOverrides={{ bg: 'azure' }}>
            <button>element 1</button>
          </ColumnsItem>
          <Columns.Item>
            <button>element 2</button>
          </Columns.Item>
        </Columns>
      </ThemeProvider>
    )

    expect(getByTestId('columns')).toHaveStyleRule('background-color', 'coral')
    expect(getByTestId('item')).toHaveStyleRule('background-color', 'azure')
  })

  it('should match snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Columns>
          <ColumnsItem>
            <button>element 1</button>
          </ColumnsItem>
          <Columns.Item>
            <button>element 2</button>
          </Columns.Item>
        </Columns>
        <Columns>
          <ColumnsItem units={4}>
            <button>element 1</button>
          </ColumnsItem>
          <Columns.Item units={4} offset="right">
            <button>element 2</button>
          </Columns.Item>
        </Columns>
      </ThemeProvider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not have a11y violations', async () => {
    const { container } = render(
      <ThemeProvider>
        <Columns>
          <ColumnsItem>
            <button>element 1</button>
          </ColumnsItem>
          <Columns.Item>
            <button>element 2</button>
          </Columns.Item>
        </Columns>
      </ThemeProvider>
    )

    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
