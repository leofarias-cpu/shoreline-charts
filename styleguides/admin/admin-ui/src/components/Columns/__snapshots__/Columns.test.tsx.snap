// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns tests should match snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-flex: 1 1 0%;
  -ms-flex: 1 1 0%;
  flex: 1 1 0%;
  max-width: 100%;
}

.emotion-0:not(:first-child) {
  padding-left: 0.25rem;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-0"
    >
      <button>
        element 1
      </button>
    </div>
    <div
      class="emotion-0"
    >
      <button>
        element 2
      </button>
    </div>
  </div>
  .emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 33.33333333333333%;
}

.emotion-0:not(:first-child) {
  padding-left: 0.25rem;
}

.emotion-1 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 33.33333333333333%;
  margin-right: auto;
}

.emotion-1:not(:first-child) {
  padding-left: 0.25rem;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-0"
    >
      <button>
        element 1
      </button>
    </div>
    <div
      class="emotion-1"
    >
      <button>
        element 2
      </button>
    </div>
  </div>
</DocumentFragment>
`;
